src/Composer/InstalledVersions.php

```diff
@@ -39,14 +39,37 @@ public static function getInstalledPackages()
             $packages[] = array_keys($installed['versions']);
         }
 
-
         if (1 === \count($packages)) {
             return $packages[0];
         }
 
         return array_keys(array_flip(\call_user_func_array('array_merge', $packages)));
     }
 
+    /**
+     * Returns a list of all package names with a specific type e.g. 'library'
+     *
+     * @param  string   $type
+     * @return string[]
+     * @psalm-return list<string>
+     */
+    public static function getInstalledPackagesByType($type = 'library')
+    {
+        $packages = array();
+        foreach (self::getInstalled() as $installed) {
+            $packages[] = $installed['versions'];
+        }
+
+        $packagesByType = array();
+        foreach ($packages[0] as $key => $package) {
+            if (isset($package['type']) && $package['type'] == $type) {
+                $packagesByType[] = $key;
+            }
+        }
```

Comment:
You are ignoring everything but the first array returned by getInstalled here, which is not good in case there are multiple autoloaders loaded.

```suggestion
        $packagesByType = array();

        foreach (self::getInstalled() as $installed) {
            foreach ($installed['versions'] as $name => $package) {
                if (isset($package['type']) && $package['type'] === $type) {
                    $packagesByType[] = $name;
                }
            }
        }
```